### ISSUE TITLE
Optimise the performance of IdPs

### ISSUE THREAD

**Issue by joshuakarp** [MatrixAI/Polykey#165]
_Created by @CMCDragonkai_ 

The following discussion from !141 should be addressed:

- [ ] @CMCDragonkai started a [discussion](https://gitlab.com/MatrixAI/Engineering/Polykey/js-polykey/-/merge_requests/141#note_498393381):  (+2 comments)

    > Need to keep a note here regarding API limits and performance optimisation of the providers.
    > 
    > * https://docs.github.com/en/rest/overview/resources-in-the-rest-api#rate-limiting
    > * https://docs.github.com/en/rest/overview/resources-in-the-rest-api#conditional-requests
    > 
    > There's a 5K API request per hour limit shared across all oauth2 apps on a particualar user. We can keep track of the limits as well. So when we get throttled we understand we cannot contact the provider API. This should be presented to the user to prevent frustration.

So basically:

1. Resource caching
2. Rate limit tracking
3. Conditional requests (based on resource caching)

should all be applied where we can do so.

**Comment by CMCDragonkai** [comment]
@tegefaulkes this was done initially as part of github provider using the graphql API  for the discovery parts (gists).

So in this case, github rest API isn't being used anymore I think.

Actually can you check what parts of the provider is still REST API. Do those REST API limits still apply?

Either way this issue should be cancelled for more specific optimisations for the social discovery issues in Polykey#696 Polykey#462 Polykey#691 Polykey#692 Polykey#702 Polykey#740 Polykey#737
Polykey-CLI#162 Polykey-CLI#163 Polykey-CLI#181 Polykey-CLI#203 Polykey-CLI#204
